### PR TITLE
Use ArtifactView API in IDE plugins

### PR DESCRIPTION
**Pull request by oehme**
For a long time the IDE plugins have been using a
half-baked custom resolution engine which was slow
and buggy. Most importantly they were using APIs like
SelfResolvingDependency and ResolvedConfiguration which
we want to deprecate soon. Since these APIs don't contain
enough information, it was also using the resolved graph
to get that information. Accessing the full graph is memory
intensive though, so it should be avoided when possible.

They now use the new ArtifactView API, which is more efficient.
This also removes any potential discrepancies between resolution
during normal task execution and resolution when importing into
an IDE.

@nskvortsov you might want to give this branch a try, it should speed things up when importing into IDEA.

@bigdaz I have some follow-up changes planned that will use a single artifact view instead of querying 4 times. But I wanted to get this first step reviewed before I continue.

### CODE REVIEW

**Comment by ddimtirov** [comment]
As this is touching the related code, could you try to also address the classpath ordering problem raised [in this forum topic](https://discuss.gradle.org/t/gradle-messes-up-the-classpath-order-in-generated-projects-when-there-are-mixed-dependency-types/13130/2)?

**Comment by oehme** [comment]
@ddimtirov I have some follow-up changes planned that will get us closer to that. However, I will not change the behavior while I'm still refactoring/optimizing, as that increases the risk of missing a breaking change and makes reviews harder. Also, fixing it at this level is not enough, it would also require some changes to Buildship (and possibly IDEA). 

The only use case I'm currently aware of is to "shadow" individual classes, which is a bad idea, because it means the project depends on the order of the classpath. Note that Gradle does not guarantee any particular ordering of artifacts, it only guarantees that it is reproducible given the same inputs and same Gradle version. So even when compiling on the command line this could break at any point. Java 9 forbids this completely, so that use case has no future in any case. A much better solution is to create a full copy of the artifact you want to change and then use a resolution rule to say "When you see the original one, replace it with my adjusted version". That way you no longer depend on ordering, but always have your replaced version on the classpath.

Of course I might have missed a use case and just reducing mental dissonance between CLI and IDE could be worth it. Let me know what you think.

**Comment by ddimtirov** [comment]
@oehme My use case is indeed shadowing. While i do agree it is not a good practice, in real life it is a duct tape that have saved the day on multiple occasions. Often the upstream maintainers can not or will not accommodate our changes, and creating and maintaining alternative artifact requires extra work and governance (especially in case the upstream changes fast). The Java 9 concern is also a good point, but we'll cross that bridge when we come to it.

The artifact ordering of Gradle configurations, while not documented is intuitive topological sorting and hasn't changed since the 0.98 days. The IDE ordering is also stable, but different, as the modules are sorted first by whether a project or repo artifact, and then topologically in each group.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/IdeDependenciesExtractor.java [hunk 1]
```diff
@@ -75,7 +75,7 @@
         // can have multiple IDE dependencies with same component identifier (see GRADLE-1622)
         Multimap<ComponentIdentifier, IdeExtendedRepoFileDependency> resolvedDependenciesComponentMap = LinkedHashMultimap.create();
         for (IdeExtendedRepoFileDependency dep : resolvedExternalDependencies(plusConfigurations, minusConfigurations)) {
-            resolvedDependenciesComponentMap.put(toComponentIdentifier(dep.getId()), dep);
+            resolvedDependenciesComponentMap.put(dep.getComponentIdentifier(), dep);
```
Comment (bigdaz) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/IdeDependenciesExtractor.java]: I guess `toComponentIdentifier()` can be removed.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/model/IdeRepoFileDependency.java [hunk 2]
```diff
@@ -33,10 +36,20 @@ public File getFile() {
     }
 
     public ModuleVersionIdentifier getId() {
+        if (componentIdentifier == null) {
+            return null;
+        }
+        if (id == null) {
+            id = new DefaultModuleVersionIdentifier(componentIdentifier.getGroup(), componentIdentifier.getModule(), componentIdentifier.getVersion());
+        }
```
Comment (bigdaz) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/model/IdeRepoFileDependency.java]: Could use `DefaultModuleVersionIdentifier.newId(ModuleComponentIdentifier)`. And perhaps we should make that null-safe.

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/IdeDependenciesExtractor.java (hunk 1 above)
Comment (oehme) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/IdeDependenciesExtractor.java]: Yeah, that dies in the next PR :)

**Pull request merged**